crypto/x509/x509_vfy.c

```diff
@@ -2980,7 +2984,14 @@ static int build_chain(X509_STORE_CTX *ctx)
                  * certificate among the ones from the trust store.
                  */
                 if ((search & S_DOALTERNATE) != 0) {
-                    OPENSSL_assert(num > i && i > 0 && ss == 0);
+                    if (!ossl_assert(num > i && i > 0 && ss == 0)) {
+                        X509err(X509_F_BUILD_CHAIN, ERR_R_INTERNAL_ERROR);
+                        X509_free(xtmp);
+                        trust = X509_TRUST_REJECTED;
+                        ctx->error = X509_V_ERR_UNSPECIFIED;
+                        search = 0;
+                        continue;
```

Comment:
> WRT /* last change to make a trusted chain */, I don't think it can be moved into the switch. That block updates the value of trust which is then used by the subsequent switch.

Yes, but *how*? It's trivial logic with return 1 or 0. And it can as well be done in case X509_TRUST_UNTRUSTED. Basically what originally stung the eye is code duplication. Indeed both "/* last chance */" and case X509_TRUST_UNTRUSTED call num = sk_X509_num(ctx->chain); followed by if (num <= depth) or equivalent if (num > depth) {} else... And this makes you wonder if  code would be more readable if there was no duplication.

As for concern about control flow and suitability of making sk_X509_num call. Yeah, it seems to be ungrounded.